### PR TITLE
npm refactoring

### PR DESCRIPTION
**Pull request by jmlemetayer**
The patchset refactors the npm support in Yocto:
 - The npm fetcher downloads only the package source (and not the dependencies,
   like the other fetchers) (patch submitted in the bitbake-devel list).
 - The npm class handles the dependencies with npm (and not manually).
 - The npm recipe creation is simplified to avoid issues.

This patchset may remove some features (lockdown file, license management for
dependencies) but fixes the majority of the npm issues. All of these issues
from the bugzilla.yoctoproject.org are resolved by this patchset:
#10237, #10760, #11028, #11728, #11902, #12534

The fetcher and recipetool are now aware of a 'latest' keyword for the version
which allow to build the latest version available on the registry. This feature
fixes the two issues: #10515, #11029

Moreover, even if I cannot test the issue #13415, I think it should be fixed.

I have tested the recipe creation and builds using a self made example made to
generate build failures. It can be found here:
 - https://github.com/savoirfairelinux/node-server-example
 - https://npmjs.com/package/@savoirfairelinux/node-server-example

The test steps are these ones:
  $ source poky/oe-init-build-env
  $ bitbake-layers add-layer ../meta-openembedded/meta-oe
  $ devtool add "npm://registry.npmjs.org;name=@savoirfairelinux/node-server-example;version=latest"
  $ devtool build savoirfairelinux-node-server-example
  $ bitbake-layers create-layer ../meta-test
  $ bitbake-layers add-layer ../meta-test
  $ devtool finish savoirfairelinux-node-server-example ../meta-test
  $ echo IMAGE_INSTALL_append = '" savoirfairelinux-node-server-example"' >> conf/local.conf
  $ bitbake core-image-minimal

And the 'devtool add' urls could be one of these:
 - npm://registry.npmjs.org;name=@savoirfairelinux/node-server-example;version=latest
 - npm://registry.npmjs.org;name=@savoirfairelinux/node-server-example;version=1.0.0
 - git://github.com/savoirfairelinux/node-server-example.git;protocol=https

I also tested with angular applications with huge dependencies and it worked
very well. When this patchset will be merged I will be able to submit a full
angular application support in Yocto.

### CODE REVIEW

File: bitbake/lib/bb/fetch2/npm.py [hunk 1]
```diff
@@ -2,300 +2,200 @@
...
-    def init(self, d):
-        pass
+    """
+        Class to fetch a package from a npm registry
+    """
 
     def supports(self, ud, d):
         """
-        Check to see if a given url can be fetched with npm
+            Check to see if a given url can be fetched with npm
```
Comment (gdoffe) [inline bitbake/lib/bb/fetch2/npm.py]: remove 'to see'

File: bitbake/lib/bb/fetch2/npm.py [hunk 2]
```diff
@@ -2,300 +2,200 @@
...
-        logger.debug(2, "Calling cleanup %s" % ud.pkgname)
-        bb.utils.remove(ud.localpath, False)
-        bb.utils.remove(ud.pkgdatadir, True)
-        bb.utils.remove(ud.fullmirror, False)
+        return ud.type in ["npm"]
 
     def urldata_init(self, ud, d):
         """
-        init NPM specific variable within url data
+            Init npm specific variable within url data
```
Comment (gdoffe) [inline bitbake/lib/bb/fetch2/npm.py]: variables ?

File: bitbake/lib/bb/fetch2/npm.py [hunk 3]
```diff
@@ -2,300 +2,200 @@
...
+        """
+
+        # Note that when using a version which does not exist on the registry
+        # (like 'version=latest') the local filename will use this version
+        # string (e.g 'my-package-latest.tgz') but the actual downloaded file
+        # will be the latest version available on the registry (e.g '1.2.3').
+        #
+        # This trick can be useful with devtool / recipetool to automatically
+        # fetch the latest version, BUT these files (the ones with non-existent
+        # versions) must be updated every times.
```
Comment (gdoffe) [inline bitbake/lib/bb/fetch2/npm.py]: If I understand well, it is not  related to 'latest' keyword it could be anything else ?
If its the case it is dangerous, the described behavior should only be applied in case of 'latest' version specified.

File: bitbake/lib/bb/fetch2/npm.py [hunk 4]
```diff
@@ -2,300 +2,200 @@
...
+
         return True
 
-    def _runpack(self, ud, d, pkgfullname: str, quiet=False) -> str:
+    @staticmethod
+    def _run_wget(ud, d, cmd):
         """
-        Runs npm pack on a full package name.
-        Returns the filename of the downloaded package
+            Run the 'wget' command with progress
```
Comment (gdoffe) [inline bitbake/lib/bb/fetch2/npm.py]: 'download progress status' you mean ?

File: meta/classes/npm.bbclass [hunk 5]
```diff
@@ -1,94 +1,171 @@
...
-NPM_INSTALL_DEV ?= "0"
+NPM_ARCH ?= "${@npm_arch(d.getVar('TARGET_ARCH'))}"
+
+NPM_INSTALL_EXTRA_ARGS ?= ""
+
+B = "${WORKDIR}/build"
+
+npm_install_shrinkwrap() {
+    # This function ensures that there is a shrinkwrap file in the specified
+    # directory. A shrinkwrap file is mandatory to be able to have reproducible
```
Comment (gdoffe) [inline meta/classes/npm.bbclass]: remove 'to be able'

File: meta/classes/npm.bbclass [hunk 6]
```diff
@@ -1,94 +1,171 @@
...
+    # builds. If the shrinkwrap file is not already included in the sources,
+    # the recipe can provides one by using the NPM_SHRINKWRAP option.
```
Comment (gdoffe) [inline meta/classes/npm.bbclass]: provide

File: meta/classes/npm.bbclass [hunk 7]
```diff
@@ -1,94 +1,171 @@
...
+    install -d ${D}/${libdir}
+    cp --no-preserve=ownership --recursive ${B}/lib/. ${D}/${libdir}
+
+    if [ -d "${B}/bin" ]
+    then
+        install -d ${D}/${bindir}
+        cp --no-preserve=ownership --recursive ${B}/bin/. ${D}/${bindir}
+    fi
+
+    # If the package (or his dependencies) uses node-gyp to build native addons,
```
Comment (gdoffe) [inline meta/classes/npm.bbclass]: 'its' not 'his'

**Review by gdoffe — CHANGES_REQUESTED** [review]
**devtool/standard.py: update the append file for the npm recipes:**
What is the 'append file' and in which context is this file used ?
I guess it is related to an yocto recipe, but it could be useful to mention it.

**Review by gdoffe — COMMENTED** [review]
**fetch2/npm.py: refactor the npm fetcher:**
That's a big commit with new features, bug fixes, etc...
Is this possible to split thta commit in something more atomic and precise ?

**Review by gdoffe — COMMENTED** [review]
**npm.bbclass: refactor the npm class:**
I would replace:
"The simplest solution is to keep it as simple as possible and to let
npm do its job."
by:
"The simplest solution is to let npm do its job."

**Review by gdoffe — CHANGES_REQUESTED** [review]
**recipetool/create_npm.py: refactor the npm recipe creation handler:**
generates

File: bitbake/lib/bb/fetch2/npm.py (hunk 3 above)
Comment (jmlemetayer) [reply bitbake/lib/bb/fetch2/npm.py]: Your right. I will update the behavior to accept only existent version or the "latest" keyword.

**Comment by jmlemetayer** [comment]
> **devtool/standard.py: update the append file for the npm recipes:**
> What is the 'append file' and in which context is this file used ?
> I guess it is related to an yocto recipe, but it could be useful to mention it.

Done

**Comment by jmlemetayer** [comment]
> The simplest solution is to let npm do its job

Done

**Comment by jmlemetayer** [comment]
> **recipetool/create_npm.py: refactor the npm recipe creation handler:**
> generates

Done

**Comment by jmlemetayer** [comment]
> **fetch2/npm.py: refactor the npm fetcher:**
> That's a big commit with new features, bug fixes, etc...
> Is this possible to split thta commit in something more atomic and precise ?

The refactoring is done on 3 files: the fetcher, the class and the recipetool handler. This is very hard to split the diff on these files.
However I split mandatory diff on the other files as atomic as possible.

File: meta/classes/npm.bbclass [hunk 8]
```diff
@@ -1,94 +1,171 @@
...
 
-NPMPN ?= "${@node_pkgname(d)}"
+NPM_INSTALL_DEV ?= "0"
 
-NPM_INSTALLDIR = "${libdir}/node_modules/${NPMPN}"
+NPM_REGISTRY ?= "https://registry.npmjs.org"
 
-# function maps arch names to npm arch names
-def npm_oe_arch_map(target_arch, d):
+def npm_arch(target_arch):
```
Comment (eloi29) [inline meta/classes/npm.bbclass]: You are renaming the function and removing the function description.
I would do that in a seperate commit

File: meta/classes/npm.bbclass [hunk 9]
```diff
@@ -1,94 +1,171 @@
...
-            if description:
-                d.setVar('SUMMARY_%s' % expanded_pkgname, description.replace(u"\u2018", "'").replace(u"\u2019", "'"))
-    d.appendVar('RDEPENDS_%s' % d.getVar('PN'), ' %s' % ' '.join(pkgnames).replace('_', '-'))
+    # node(1) is using /usr/lib/node as default include directory and npm(1) is
+    # using /usr/lib/node_modules as install directory. Let's make both happy.
+    ln -fs node_modules ${D}/${libdir}/node
 }
 
-FILES_${PN} += " \
+FILES_${PN} = " \
```
Comment (eloi29) [inline meta/classes/npm.bbclass]: I would explain in a separate commit why you are using 
`FILES_${PN} = " ` instead of `FILES_${PN} += " `

File: meta/classes/npm.bbclass [hunk 10]
```diff
@@ -1,94 +1,171 @@
...
     ${bindir} \
-    ${libdir}/node \
-    ${NPM_INSTALLDIR} \
+    ${libdir} \
```
Comment (eloi29) [inline meta/classes/npm.bbclass:167]: I would explain in a separate commit why:
- renaming ${libdir}/node 
- removing ${NPM_INSTALLDIR}

File: meta/classes/npm.bbclass [hunk 11]
```diff
@@ -1,94 +1,171 @@
...
+    # This function executes the 'npm install' command which builds and
+    # installs every dependencies needed for the package. All the files are
+    # installed in a build directory ${B} without filtering anything. To do so,
+    # a combination of 'npm pack' and 'npm install' is used to ensure that the
+    # files in ${B} are actual copies instead of symbolic links (which is the
+    # default npm behavior).
+
+    # The npm command use by default a cache which is located in '~/.npm'. In
+    # order to force the next npm commands to disable caching, the npm cache
+    # needs to be cleared. But to not mess with the local cache, the npm
```
Comment (eloi29) [inline meta/classes/npm.bbclass]: `but not to mess instead `of `but to not mess`

Comment (eloi29) [reply meta/classes/npm.bbclass]: In addition, the verb "mess" looks to be informal. I would use an other word

File: meta/classes/npm.bbclass [hunk 12]
```diff
@@ -1,94 +1,171 @@
...
+    # config needs to be updated to use another cache directory. And to not
+    # mess with the local '~/.npmrc` file, the HOME needs to be updated as well.
```
Comment (eloi29) [inline meta/classes/npm.bbclass]: I would reword like below:

the HOME needs to be updated as well to avoid conflicts with the local '~/.npmrc' file.

**Review by eloi29 — CHANGES_REQUESTED** [review]

**Review by eloi29 — COMMENTED** [review]
> 
> This commit refactors the npm fetcher to improve some points and fix
> others:
I would name it. What are you improving ? What are you fixing ?

> 
> The big change is that the fetcher is only fetching the package source
> and no more the dependencies. This is to act as the other fetchers
> (git, wget, ...). The dependencies will now be fetched by the npm class.
> 

I would reword:

` This is to act as the other fetchers (git, wget, ...) ` to `Thus the npm fetcher act as the other fetchers e.g git, wget`

> This commit also fixes a lot of issues with the package names (exotic
> characters, scoped packages) which were badly handled.

Have you got an exhaustive list of fixed issues ?

> 
> The validation files - lockdown.json and npm-shrinkwrap.json - are no
> more used by the fetcher. Instead, the downloaded tarball is validated
> with the shasum provided in the npm view of the package. The shrinkwrap
> file is now used by the npm class.

Is the use of shasum the standard way outside Yocto ? If it does, mention it.

> 
> Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

**Review by eloi29 — APPROVED** [review]

File: scripts/lib/recipetool/create_npm.py [hunk 13]
```diff
@@ -1,321 +1,215 @@
...
     def process(self, srctree, classes, lines_before, lines_after, handled, extravalues):
-        import bb.utils
-        import oe.package
-        from collections import OrderedDict
+        """
+            Handle the npm recipe creation
+        """
 
-        if 'buildsystem' in handled:
+        if "buildsystem" in handled:
```
Comment (eloi29) [inline scripts/lib/recipetool/create_npm.py]: why using double quotes instead of simple quotes ?

**Review by eloi29 — CHANGES_REQUESTED** [review]

**Review by eloi29 — CHANGES_REQUESTED** [review]
> When creating a recipe using devtool, a workspace is created to store
> the new recipe, the recipe source and some append files. These append
> files are used by devtool to build the recipe using externalsrc (to use
> the source which are in the workspace). They can also have some
> additional action according to the class on the recipe.

I would reword `They can also have some additional action according to the class on the recipe.`
I don't understand what you mean.
Otherwise LGTM.
> 
> This commit updates the append file for the npm recipes. The
> devtool / externalsrc files are removed in the npm build directory
>

**Review by eloi29 — APPROVED** [review]

**Review by eloi29 — CHANGES_REQUESTED** [review]
I would write `the "noverify" parameter` instead of `the noverify parameter`.

**Comment by jmlemetayer** [comment]
> > This commit refactors the npm fetcher to improve some points and fix
> > others:
> > I would name it. What are you improving ? What are you fixing ?
> 
> > The big change is that the fetcher is only fetching the package source
> > and no more the dependencies. This is to act as the other fetchers
> > (git, wget, ...). The dependencies will now be fetched by the npm class.
> 
> I would reword:
> 
> `This is to act as the other fetchers (git, wget, ...)` to `Thus the npm fetcher act as the other fetchers e.g git, wget`
> 
> > This commit also fixes a lot of issues with the package names (exotic
> > characters, scoped packages) which were badly handled.
> 
> Have you got an exhaustive list of fixed issues ?
> 
> > The validation files - lockdown.json and npm-shrinkwrap.json - are no
> > more used by the fetcher. Instead, the downloaded tarball is validated
> > with the shasum provided in the npm view of the package. The shrinkwrap
> > file is now used by the npm class.
> 
> Is the use of shasum the standard way outside Yocto ? If it does, mention it.
> 
> > Signed-off-by: Jean-Marie LEMETAYER [jean-marie.lemetayer@savoirfairelinux.com](mailto:jean-marie.lemetayer@savoirfairelinux.com)

Done

**Comment by jmlemetayer** [comment]
> > When creating a recipe using devtool, a workspace is created to store
> > the new recipe, the recipe source and some append files. These append
> > files are used by devtool to build the recipe using externalsrc (to use
> > the source which are in the workspace). They can also have some
> > additional action according to the class on the recipe.
> 
> I would reword `They can also have some additional action according to the class on the recipe.`
> I don't understand what you mean.
> Otherwise LGTM.
> 
> > This commit updates the append file for the npm recipes. The
> > devtool / externalsrc files are removed in the npm build directory

Done

**Comment by jmlemetayer** [comment]
> I would write `the "noverify" parameter` instead of `the noverify parameter`.

Done
